File: packages/icoder-sdk/src/resources/billing.ts
// Billing & Usage resource
import type { AxiosInstance } from 'axios';
import type { UsageSummary } from '../types.js';
import { AsyncPageNumberPager } from '../pagination.js';
import { requestConfig, type iCoDerRequestOptions } from '../request-options.js';

export interface BillingTransaction {
  id: string;
  date: string;
  description: string;
  amount: string;
  type: string;
  source?: string | null;
  balance_after?: number | null;
}

export interface BillingTransactionPage {
  transactions: BillingTransaction[];
  total: number;
  page: number;
  page_size: number;
}

export interface BillingRunSettlement {
  run_id: string;
  status: string;
  reserved_amount: number;
  settled_amount: number;
  currency: string;
  error_code?: string | null;
  created_at?: string | null;
}

export interface BillingRunSettlementPage {
  items: BillingRunSettlement[];
  total: number;
  page: number;
  page_size: number;
  simulation: boolean;
}

function positiveInteger(value: number, name: string, maximum?: number): number {
  if (!Number.isInteger(value) || value < 1 || (maximum !== undefined && value > maximum)) {
    throw new RangeError(`${name} must be a positive integer${maximum ? ` at most ${maximum}` : ''}`);
  }
  return value;
}

export class BillingResource {
  constructor(private http: AxiosInstance) {}

  async balance(options?: iCoDerRequestOptions): Promise<{
    balance: number;
    reserved?: number;
    available?: number;
    currency: string;
    simulation?: boolean;
    ledger_authoritative?: boolean;
    quota?: { kind: string; limit: number | null; remaining: number; enforced: boolean };
    alerts?: { low_balance: boolean; threshold: number };
  }> {
    const { data } = await this.http.get('/api/billing/balance', requestConfig(options));
    return data;
  }

  async transactions(
    page = 1,
    pageSize = 20,
    options?: iCoDerRequestOptions,
  ): Promise<BillingTransactionPage> {
    positiveInteger(page, 'page');
    positiveInteger(pageSize, 'pageSize', 100);
    const params = { page, page_size: pageSize };
    const { data } = await this.http.get<BillingTransactionPage>(
      '/api/billing/transactions',
      requestConfig(options, params),
    );
    return data;
  }

  /** Lazily iterate transaction history without pre-fetching or unbounded loops. */
  iterateTransactions(
    pageSize = 20,
    options?: iCoDerRequestOptions,
    pagerOptions: { initialPage?: number; maxPages?: number } = {},
  ): AsyncPageNumberPager<BillingTransactionPage, BillingTransaction> {
    positiveInteger(pageSize, 'pageSize', 100);
    return new AsyncPageNumberPager(
      (page) => this.transactions(page, pageSize, options),
      { items: (response) => response.transactions, totalItems: (response) => response.total },
      pagerOptions,
    );
  }

  /** Development-only ledger simulation; cloud mode rejects this mutation. */
  async simulateDebit(amount: number, reference: string, options?: iCoDerRequestOptions): Promise<{
    status: string;
    debited: number;
    new_balance: number;
    simulation: boolean;
  }> {
    const { data } = await this.http.post(
      '/api/billing/simulation/debit',
      { amount, reference },
      requestConfig(options),
    );
    return data;
  }

  /** PHI-free, development-only Agent Run settlement audit records. */
  async runSettlements(limit = 20, options?: iCoDerRequestOptions): Promise<BillingRunSettlementPage> {
    positiveInteger(limit, 'limit', 100);
    const { data } = await this.http.get<BillingRunSettlementPage>(
      '/api/billing/run-settlements',
      requestConfig(options, { limit }),
    );
    return data;
  }

  async runSettlementPage(
    page = 1,
    pageSize = 20,
    options?: iCoDerRequestOptions,
  ): Promise<BillingRunSettlementPage> {
    positiveInteger(page, 'page');
    positiveInteger(pageSize, 'pageSize', 100);
    const params = { page, page_size: pageSize };
    const { data } = await this.http.get<BillingRunSettlementPage>(
      '/api/billing/run-settlements',
      requestConfig(options, params),
    );
    return data;
  }

  iterateRunSettlements(
    pageSize = 20,
    options?: iCoDerRequestOptions,
    pagerOptions: { initialPage?: number; maxPages?: number } = {},
  ): AsyncPageNumberPager<BillingRunSettlementPage, BillingRunSettlement> {
    positiveInteger(pageSize, 'pageSize', 100);
    return new AsyncPageNumberPager(
      (page) => this.runSettlementPage(page, pageSize, options),
      { items: (response) => response.items, totalItems: (response) => response.total },
      pagerOptions,
    );
  }

  /** Retry an idempotent failed local settlement after adding credits. */
  async retryRunSettlement(
    runId: string,
    options?: iCoDerRequestOptions,
  ): Promise<Record<string, unknown>> {
    const { data } = await this.http.post(
      `/api/billing/run-settlements/${encodeURIComponent(runId)}/retry`,
      undefined,
      requestConfig(options),
    );
    return data;
  }

  /** Reconcile old crash-orphaned reservations; active runs are skipped. */
  async reconcileStaleRunSettlements(
    olderThanSeconds = 3600,
    options?: iCoDerRequestOptions,
  ): Promise<{
    simulation: boolean;
    released: number;
    marked_retryable: number;
    skipped_active: number;
    inspected: number;
    older_than_seconds: number;
  }> {
    const { data } = await this.http.post(
      '/api/billing/run-settlements/reconcile-stale',
      undefined,
      requestConfig(options, { older_than_seconds: olderThanSeconds }),
    );
    return data;
  }
}

export class UsageResource {
  constructor(private http: AxiosInstance) {}

  async summary(days = 30, options?: iCoDerRequestOptions): Promise<UsageSummary> {
    const { data } = await this.http.get(
      '/api/usage/summary',
      requestConfig(options, { days }),
    );
    return data;
  }

  async history(days = 30, options?: iCoDerRequestOptions): Promise<any> {
    const { data } = await this.http.get(
      '/api/usage/history',
      requestConfig(options, { days }),
    );
    return data;
  }
}

File: packages/icoder-python/icoder_sdk/resources/billing.py
"""Billing and usage resources with bounded request controls and pagination."""

from __future__ import annotations

from typing import Optional

from ..client import iCoDerClient
from ..pagination import PageNumberPager
from ..request_options import RequestOptions
from ..types import UsageSummary


def _positive_integer(value: int, name: str, maximum: Optional[int] = None) -> int:
    if (
        not isinstance(value, int)
        or isinstance(value, bool)
        or value < 1
        or (maximum is not None and value > maximum)
    ):
        suffix = f" at most {maximum}" if maximum is not None else ""
        raise ValueError(f"{name} must be a positive integer{suffix}")
    return value


class BillingResource:
    def __init__(self, client: iCoDerClient):
        self._client = client

    def balance(self, request_options: Optional[RequestOptions] = None) -> dict:
        resp = self._client.get(
            "/api/billing/balance", request_options=request_options,
        )
        resp.raise_for_status()
        return resp.json()

    def transactions(
        self,
        page: int = 1,
        page_size: int = 20,
        request_options: Optional[RequestOptions] = None,
    ) -> dict:
        _positive_integer(page, "page")
        _positive_integer(page_size, "page_size", 100)
        resp = self._client.get(
            "/api/billing/transactions",
            params={"page": page, "page_size": page_size},
            request_options=request_options,
        )
        resp.raise_for_status()
        return resp.json()

    def iterate_transactions(
        self,
        page_size: int = 20,
        request_options: Optional[RequestOptions] = None,
        *,
        initial_page: int = 1,
        max_pages: int = 10000,
    ) -> PageNumberPager[dict, dict]:
        _positive_integer(page_size, "page_size", 100)
        return PageNumberPager(
            lambda page: self.transactions(page, page_size, request_options),
            lambda response: response["transactions"],
            lambda response: response["total"],
            initial_page=initial_page,
            max_pages=max_pages,
        )

    def simulate_debit(
        self,
        amount: float,
        reference: str,
        request_options: Optional[RequestOptions] = None,
    ) -> dict:
        """Debit the local development ledger; cloud mode rejects this call."""
        resp = self._client.post(
            "/api/billing/simulation/debit",
            json={"amount": amount, "reference": reference},
            request_options=request_options,
        )
        resp.raise_for_status()
        return resp.json()

    def run_settlements(
        self,
        limit: int = 20,
        request_options: Optional[RequestOptions] = None,
    ) -> dict:
        """List PHI-free local Agent Run preauthorization settlements."""
        _positive_integer(limit, "limit", 100)
        resp = self._client.get(
            "/api/billing/run-settlements", params={"limit": limit},
            request_options=request_options,
        )
        resp.raise_for_status()
        return resp.json()

    def run_settlement_page(
        self,
        page: int = 1,
        page_size: int = 20,
        request_options: Optional[RequestOptions] = None,
    ) -> dict:
        _positive_integer(page, "page")
        _positive_integer(page_size, "page_size", 100)
        resp = self._client.get(
            "/api/billing/run-settlements",
            params={"page": page, "page_size": page_size},
            request_options=request_options,
        )
        resp.raise_for_status()
        return resp.json()

    def iterate_run_settlements(
        self,
        page_size: int = 20,
        request_options: Optional[RequestOptions] = None,
        *,
        initial_page: int = 1,
        max_pages: int = 10000,
    ) -> PageNumberPager[dict, dict]:
        _positive_integer(page_size, "page_size", 100)
        return PageNumberPager(
            lambda page: self.run_settlement_page(page, page_size, request_options),
            lambda response: response["items"],
            lambda response: response["total"],
            initial_page=initial_page,
            max_pages=max_pages,
        )

    def retry_run_settlement(
        self,
        run_id: str,
        request_options: Optional[RequestOptions] = None,
    ) -> dict:
        """Retry a failed local settlement after adding credits."""
        from urllib.parse import quote

        resp = self._client.post(
            f"/api/billing/run-settlements/{quote(run_id, safe='')}/retry",
            request_options=request_options,
        )
        resp.raise_for_status()
        return resp.json()

    def reconcile_stale_run_settlements(
        self,
        older_than_seconds: int = 3600,
        request_options: Optional[RequestOptions] = None,
    ) -> dict:
        """Reconcile old crash-orphaned reservations; active runs are skipped."""
        resp = self._client.post(
            "/api/billing/run-settlements/reconcile-stale",
            params={"older_than_seconds": older_than_seconds},
            request_options=request_options,
        )
        resp.raise_for_status()
        return resp.json()


class UsageResource:
    def __init__(self, client: iCoDerClient):
        self._client = client

    def summary(
        self,
        days: int = 30,
        request_options: Optional[RequestOptions] = None,
    ) -> UsageSummary:
        resp = self._client.get(
            "/api/usage/summary",
            params={"days": days},
            request_options=request_options,
        )
        resp.raise_for_status()
        data = resp.json()
        return UsageSummary(
            total_requests=data.get("total_requests", 0),
            credits_used=data.get("credits_used", 0),
            avg_response_time_ms=data.get("avg_response_time_ms", 0),
        )

    def history(
        self,
        days: int = 30,
        request_options: Optional[RequestOptions] = None,
    ) -> dict:
        resp = self._client.get(
            "/api/usage/history",
            params={"days": days},
            request_options=request_options,
        )
        resp.raise_for_status()
        return resp.json()
